src/Illuminate/Cache/DatabaseStore.php

```diff
@@ -149,23 +145,32 @@ public function decrement($key, $value = 1)
      * @param  string  $key
      * @param  mixed  $value
      * @param  \Closure  $callback
-     * @return void
+     * @return int|bool
      */
     protected function incrementOrDecrement($key, $value, Closure $callback)
     {
-        $prefixed = $this->prefix.$key;
+        return $this->connection->transaction(function () use ($key, $value, $callback) {
+            $prefixed = $this->prefix.$key;
 
-        $cache = $this->table()->where('key', $prefixed)->lockForUpdate()->first();
+            $cache = $this->table()->where('key', $prefixed)->lockForUpdate()->first();
+
```

Comment:
I know that MySQL will prevent that from happening by using the updated value:
> SELECT ... LOCK IN SHARE MODE sets a shared mode lock on any rows that are read. Other sessions can read the rows, but cannot modify them until your transaction commits. If any of these rows were changed by another transaction that has not yet committed, your query waits until that transaction ends and then uses the latest values.

You can reference that [here](http://dev.mysql.com/doc/refman/5.5/en/innodb-locking-reads.html). I'm not sure about the other database types though.